boot/bootutil/src/loader.c

```diff
@@ -793,58 +855,122 @@ boot_copy_sector(int flash_area_id_src, int flash_area_id_dst,
 static void
 boot_swap_sectors(int idx, uint32_t sz, struct boot_status *bs)
 {
+    const struct flash_area *fap;
     uint32_t copy_sz;
+    uint32_t trailer_sz;
     uint32_t img_off;
+    uint32_t scratch_trailer_off;
+    struct boot_swap_state swap_state;
     int rc;
 
     /* Calculate offset from start of image area. */
     img_off = boot_data.imgs[0].sectors[idx].fa_off -
               boot_data.imgs[0].sectors[0].fa_off;
 
+    copy_sz = sz;
+
+    trailer_sz = boot_slots_trailer_sz(boot_data.write_sz);
+    if (boot_data.imgs[0].sectors[idx].fa_off + sz >
+        boot_data.imgs[0].sectors[boot_data.imgs[0].num_sectors - 1].fa_off) {
+        copy_sz -= trailer_sz;
+    }
+
+    bs->use_scratch = (bs->idx == 0 && copy_sz != sz);
+
     if (bs->state == 0) {
         rc = boot_erase_sector(FLASH_AREA_IMAGE_SCRATCH, 0, sz);
         assert(rc == 0);
 
         rc = boot_copy_sector(FLASH_AREA_IMAGE_1, FLASH_AREA_IMAGE_SCRATCH,
-                              img_off, 0, sz);
+                              img_off, 0, copy_sz);
         assert(rc == 0);
 
+        if (bs->idx == 0) {
```

Comment:
> In general, I was surprised to learn this week that mcuboot copies around the image trailers as it swaps.

This was the case before but now there's no more copying of trailer. Trailer is always created (possibly using parts of an existing trailer) by `boot_status_init_by_id`.

> Has any thought been given to having a bootloader state partition that is not copied around during image swap? It seems like it would help avoid some of the issues.

I have not given it much thought because I started on the project just one month ago and was basically trying to fix these issues that were found by doing as few changes as possible. The code ended up being quite complicated. I agree that using a dedicated sector for at least the "status bytes" trailer stuff would be a really good idea. As long as we have a spare sector which hopefully will be the case.

> (By the way, the commit logs in this series are a little terse -- do you have any pointers to more detailed discussion explaining these changes? It would help me understand what's going on a bit better.)

Apart from reading the `design.txt` document, I have not. I'll try to get more background on the idea design when I'm at Runtime later this week.